README.md
# Raspberry Pi Rust Cross-Compilation Tools

## Background

A cross-compilation environment for Raspberry Pi (Arm v7) designed to work with
[rust-embedded/cross](https://github.com/rust-embedded/cross) as maintained by the
[Rust Embedded Devices WG](https://github.com/rust-embedded/wg).

This project specializes the official ARM v7 Docker image to add libraries that are
useful for developing software for the Raspberry Pi.

## DEPRECATION

The whole purpose of this project was to provide a custom Docker image that `cross`
could then invoke during the cross-compilation for Raspberry Pi targets. However,
from `cross` [V0.2.2](https://github.com/cross-rs/cross/blob/main/CHANGELOG.md#v022---2022-06-24),
it has natively supported the [`target.{{TARGET}}.image`](https://github.com/cross-rs/cross/blob/main/README.md#dockerfiles)
key that means the specification of the necessary tooling can live within the project
that needs the custom Docker image.

Since each project probably has unique requirements that seems a much better pattern to
follow, rather than having to keep two projects in sync.

To import the scripts from this project into a Rust project using `cross` for the
Raspberry Pi, copy the `docker` folder into your project and then add the following to
your project's `cross.toml`:

``` toml
[target.armv7-unknown-linux-gnueabihf]
dockerfile = "./docker/Dockerfile.armv7-unknown-linux-gnueabihf.rpi"
```

There are no further plans to update this project beyond the current version.

## Supported targets

* armv7-unknown-linux-gnueabihf

  ARM v7 - second generation Raspberry Pi and later.

Currently only generates a container for executables dynamically linked against
[GNU libc](https://www.gnu.org/software/libc/).  Further work would be necessary to
generate the libraries needed for executables statically linked against
[MUSL libc](https://musl.libc.org/about.html).

## Supported libraries

The image currently adds support to the base image for:

* [RTIMULib as distributed with Raspbian](https://github.com/RPi-Distro/RTIMULib)
  
  Required to support the [Sense HAT](https://www.raspberrypi.org/products/sense-hat/)
  using, for example, the [sensehat-rs](https://crates.io/crates/sensehat) crate.

* Python 3.x

  Required to support compiling code that will use [PyO3](https://pyo3.rs/).

## Build instructions

### Interactive debugging

Uses [floki](https://metaswitch.github.io/floki/) so just invoke `floki`, which will
build the currently defined Docker image and launch an interactive shell in the
resulting container:

```console
$ cd docker
$ floki
Building ... 
<...snip...>
Specialized Rust Cross Compiler for Raspberry Pi (ARM v7) targets
# 
```

### Manual

Use the build script, specifying the semver tag (e.g. "0.1.2"):

```console
$ cd docker
$ ./build-cross-container-no-upload.sh <SEMVER VERSION>
Create solimike/cross-amd64 build container
        version: <SEMVER VERSION>
   architecture: amd64
  PARALLEL_MAKE: -j4
[+] Building 0.2s (10/10) FINISHED           
$ 
```

And then once you're happy with the operation of the container, push to Docker Hub:

```console
$ ./push-docker-image.sh
Password: 
Login Succeeded
The push refers to repository [docker.io/solimike/raspberry_cross]
ae61c6a2b875: Preparing 
<...snip...>
armv7-unknown-linux-gnueabihf-0.0.1: digest: sha256:7897e808af24e1dfec697ac27fb0d8d584bf7a4aa9ec93c5c65ea99c4ade186c size: 4922
$
```

### CI

TODO


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

## [0.2.1]

### Changed

- Updated to latest released `cross` (V0.2.5).
- Updated to latest vesrion of Python V3.9 (V3.9.17).
- **DEPRECATED THE PROJECT** - see [why](README.md#DEPRECATION)

## [0.2.0]

### Added

- Support for RaspberryPi OS `bullseye`.

### Changed

- Updated to latest released `cross` (V0.2.4).

### Fixed

## [0.1.0]

### Added

- OpenSSL support for native Python.
- OpenSSL support for cross compiled Python.

### Changed

- Python updated to V3.7.12 to get some improved behaviours running in containers.

### Fixed

## [0.0.2]

### Added

- Python V3.7.3 for [PyO3](https://pyo3.rs/) cross compilation.

### Fixed

- Nothing

## [0.0.1] - 2021-04-04

### Added

- Initial version.

_Note that these links are currently placeholders only and do not point to this_
_codebase's changes!_

_TODO: update to reflect this codebase once published to GitHub!_

[unreleased]: https://github.com/solimike/raspberry-cross/compare/v0.2.1...HEAD
[0.2.1]: https://github.com/solimike/raspberry-cross/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/solimike/raspberry-cross/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/solimike/raspberry-cross/compare/v0.0.2...v0.1.0
[0.0.2]: https://github.com/solimike/raspberry-cross/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/solimike/raspberry-cross/releases/tag/v0.0.1


docker/rtimulib.sh
#!/usr/bin/env bash
#
# Cross compile the Raspberry Pi distro of RTIMULib.  For further background, see
# https://github.com/RPi-Distro/RTIMULib
#
# This script draws heavily on the conventions of the Embedded Rust cross-compilation
# tools - see for example scripts such as:
#
#     https://github.com/rust-embedded/cross/blob/master/docker/musl.sh

set -x
set -euo pipefail

hide_output() {
    set +x
    trap "
      echo 'ERROR: An error was encountered with the build.'
      cat /tmp/build.log
      exit 1
    " ERR
    bash -c 'while true; do sleep 30; echo $(date) - building ...; done' &
    PING_LOOP_PID=$!
    "${@}" &> /tmp/build.log
    trap - ERR
    kill "${PING_LOOP_PID}"
    set -x
}

main() {
    # Version of RTIMULib to build.
    local version=7.2.1

    # Ensure that we satisfy any package dependencies we need in order to build.
    local dependencies=(
        ca-certificates
        curl
        build-essential
    )

    apt-get update
    local purge_list=()
    for dep in "${dependencies[@]}"; do
        if ! dpkg -L "${dep}"; then
            apt-get install --assume-yes --no-install-recommends "${dep}"
            purge_list+=( "${dep}" )
        fi
    done

    # Create a temp directory for the build.
    local td
    td="$(mktemp -d)"
    echo "Building in ${td}"

    # Fetch the source tarball.
    pushd "${td}"
    curl --retry 3 -sSfL "https://github.com/RPi-Distro/RTIMULib/archive/V${version}.tar.gz" -O
    tar --strip-components=1 -xzf "V${version}.tar.gz"

    # Create a cross compilation build environment.
    toolchain=${1}
    echo "Cross compiling using ${toolchain} toolchain."
    mkdir -p Linux/build 
    cd Linux/build
    hide_output cmake -DCMAKE_TOOLCHAIN_FILE=/${toolchain} \
          -DBUILD_GL=OFF \
          -DBUILD_DEMOGL=OFF .. 

    # Build & install the library.
    hide_output cmake --build . 
    hide_output cmake --install .
    
    # Remove any build packages we needed.
    if (( ${#purge_list[@]} )); then
      apt-get purge --assume-yes --auto-remove "${purge_list[@]}"
    fi

    # Cleanup all the files we used.
    popd
    rm -rf "${td}"
    rm "${0}"
    rm "${toolchain}"
}

main "${@}"


docker/build_local_python.sh
#!/usr/bin/env bash
#
# Cross compile the Raspberry Pi distro of Python.  
#
# This script draws heavily on the conventions of the Embedded Rust cross-compilation
# tools - see for example scripts such as:
#
#     https://github.com/rust-embedded/cross/blob/master/docker/musl.sh
#
# First of all we build the Python for the container itself (i.e. a native build)
# then we do a cross compilation for the target ARM architecture. This is necessary
# because the cross-complilation of Python relies on a host installation of Python and
# the Python 3.5 that is currently packaged with the Ubuntu 16.04 that is the base for 
# Rusts cross compilation tools is too old to support cross compilation of later 
# versions of Python (e.g. Python 3.7) that rely on features like f-strings.
#
# Sources of info:
#
#   1)  https://gitlab.com/Spindel/rust-python-cross-example
#   2)  https://raspberrypi.stackexchange.com/questions/66528/how-to-cross-compile-python-3-6-for-the-raspberry-pi
#   3)  https://github.com/rust-embedded/cross/issues/556

set -x
set -euo pipefail

hide_output() {
    set +x
    trap "
      echo 'ERROR: An error was encountered with the build.'
      cat /tmp/build.log
      exit 1
    " ERR
    bash -c 'while true; do sleep 30; echo $(date) - building ...; done' &
    PING_LOOP_PID=$!
    "${@}" &> /tmp/build.log
    trap - ERR
    kill "${PING_LOOP_PID}"
    set -x
}

main() {
    # Version of Python to build.
  local version=${1}
    echo "Building native Python V${version}"

    # Create a temp directory for the build.
    local td
    td="$(mktemp -d)"
    echo "Building in ${td}"

    # Fetch the source tarball.
    pushd "${td}"
    curl --retry 3 -sSfL "https://www.python.org/ftp/python/${version}/Python-${version}.tgz" -O
    tar --strip-components=1 -xzf "Python-${version}.tgz"

    # Do the native build. The subsequent cross-compile needs "python" to refer
    # to the python3 executable.
    # ./configure --enable-shared
    ./configure 
    make install
    ln -s python3 /usr/local/bin/python

    # Cleanup all the files we used.
    popd
    rm -rf "${td}"
    rm "${0}"
}

main "${@}"


docker/build_cross_python.sh
#!/usr/bin/env bash
#
# Cross compile the Raspberry Pi distro of Python.  
#
# This script draws heavily on the conventions of the Embedded Rust cross-compilation
# tools - see for example scripts such as:
#
#     https://github.com/rust-embedded/cross/blob/master/docker/musl.sh
#
# Sources of info:
#
#   1)  https://gitlab.com/Spindel/rust-python-cross-example
#   2)  https://raspberrypi.stackexchange.com/questions/66528/how-to-cross-compile-python-3-6-for-the-raspberry-pi
#   3)  https://github.com/rust-embedded/cross/issues/556

set -x
set -euo pipefail

hide_output() {
    set +x
    trap "
      echo 'ERROR: An error was encountered with the build.'
      cat /tmp/build.log
      exit 1
    " ERR
    bash -c 'while true; do sleep 30; echo $(date) - building ...; done' &
    PING_LOOP_PID=$!
    "${@}" &> /tmp/build.log
    trap - ERR
    kill "${PING_LOOP_PID}"
    set -x
}

main() {
    # Version of Python to build.
    local version=${1}
    echo "Building cross compiled Python V${version}"

    # Create a temp directory for the build.
    local td
    td="$(mktemp -d)"
    pushd "${td}"
    echo "Building in ${td}"

    # Fetch ARM architecture dependencies: tell APT where to find the resources and
    # add the architecture first. The addition of armhf to the sources list is a bit
    # crude so triggers some "not found" errors in the update which we have to ignore.
    sed -i "s/i386]/i386,armhf]/" /etc/apt/sources.list
    dpkg --add-architecture armhf
    apt-get update || true
    apt-get install --assume-yes libssl-dev:armhf \
                                 libffi-dev:armhf \
                                 zlib1g-dev:armhf 

    # Fetch the source tarball.
    curl --retry 3 -sSfL "https://www.python.org/ftp/python/${version}/Python-${version}.tgz" -O
    tar --strip-components=1 -xzf "Python-${version}.tgz"

    # Configure the toolchain and then cross-compile into the alternate location.
    export CC=arm-linux-gnueabihf-gcc
    export LD=arm-linux-gnueabihf-ld
    export CXX=arm-linux-gnueabihf-g++
    export CPP=arm-linux-gnueabihf-cpp
    export READELF=arm-linux-gnueabihf-readelf
    export RANLIB=arm-linux-gnueabihf-ranlib
    export AR=arm-linux-gnueabihf-ar

    echo ac_cv_file__dev_ptmx=no > ./config.site
    echo ac_cv_file__dev_ptc=no >> ./config.site

    CONFIG_SITE=config.site ./configure --prefix=/usr/arm-linux-gnueabihf \
                                        --build=x86_64-pc-linux-gnu \
                                        --host=arm-unknown-linux-gnueabihf \
                                        --disable-ipv6 \
                                        --enable-shared
    make install

    # Cleanup all the files we used.
    popd
    rm -rf "${td}"
    rm "${0}"
}

main "${@}"
